src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerContextTests
+    {
+        public static bool IsNotWindows7 { get; } = !PlatformDetection.IsWindows7 && PlatformDetection.IsNotOneCoreUAP;
+
+        public static IEnumerable<object[]> SubProtocol_TestData()
+        {
+            yield return new object[] { new string[0], null };
+
+            yield return new object[] { new string[] { "MyProtocol1" }, null };
+            yield return new object[] { new string[] { "MyProtocol" }, "MyProtocol" };
+            yield return new object[] { new string[] { "MyProtocol" }, "myPROTOCOL" };
+
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, null };
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [MemberData(nameof(SubProtocol_TestData))]
+        public async Task AcceptWebSocketAsync_ValidSubProtocol_Success(string[] clientProtocols, string serverProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                HttpListenerWebSocketContext socketContext = context.AcceptWebSocketAsync(serverProtocol).Result;
+                Assert.Equal(serverProtocol, socketContext.WebSocket.SubProtocol);
+            }, subProtocols: clientProtocols);
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7))]
+        // The managed implementation doesn't validate that the socket is actually a web socket.
+        // This means that HttpListener can connect to a Socket masquerading as a web socket.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task AcceptWebSocketAsync_SocketSpoofingAsWebSocket_ThrowsWebSocketException()
```

Comment:
You can see a simple web socket client connection logic here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L74
It just opens a socket to the target, and sends an HTTP request with the relevant headers, then validates the response headers, at which point the web socket connection is considered opened.  If the Windows implementation is failing and the managed succeeding here, that suggests that the HttpListener validation is different between the two, and we should understand and fix the differences in the managed one.